README.md
## What is this fork?
This is a simple fork of signature_pad 2.1.1 to be compatible with angular2-signaturepad.  It's pre-compiled for simplicity as it's only a temporary fix until the authors of the original packages can come up with a solution.

This includes a bug fix to export SVGs with the correct path colors.

## Why was this forked?
angular2-signaturepad is stuck using signature_pad 1.5.3 due to the way later versions export the class.  The current version of signature_pad as of today is 2.1.1 and includes the ability to export to SVG.  In order to get the two to play together, it was required to fork both projects.

## Yay! How do I use this?

Step 1 is to remove any installed versions of these packages:

```
npm uninstall --save angular2-signaturepad
npm uninstall --save signature_pad
npm install --save https://github.com/DynamoEffects/angular2-signaturepad.git
```

In theory that should work.

## Does this come with support and a money back guarantee?

No and no, this was forked for a very specific project and is being shared to help others who might be waiting for an update for those two projects.



# Signature Pad [![npm](https://d25lcipzij17d.cloudfront.net/badge.svg?id=js&type=6&v=2.1.1&x2=0)](https://www.npmjs.com/package/signature_pad) [![Code Climate](https://codeclimate.com/github/szimek/signature_pad.png)](https://codeclimate.com/github/szimek/signature_pad)


Signature Pad is a JavaScript library for drawing smooth signatures. It's HTML5 canvas based and uses variable width Bézier curve interpolation based on [Smoother Signatures](http://corner.squareup.com/2012/07/smoother-signatures.html) post by [Square](https://squareup.com).
It works in all modern desktop and mobile browsers and doesn't depend on any external libraries.

![Example](https://f.cloud.github.com/assets/9873/268046/9ced3454-8efc-11e2-816e-a9b170a51004.png)

## Demo
[Demo](http://szimek.github.io/signature_pad) works in desktop and mobile browsers. You can check out its [source code](https://github.com/szimek/signature_pad/blob/gh-pages/js/app.js) for some tips on how to handle window resize and high DPI screens. You can also find more about the latter in [HTML5 Rocks tutorial](http://www.html5rocks.com/en/tutorials/canvas/hidpi).

## Installation
You can install the latest release using npm:
```bash
npm install --save signature_pad
```
or Yarn:
```bash
yarn add signature_pad
```

You can also add it directly to your page using `<script>` tag:
```html
<script src="https://cdnjs.cloudflare.com/ajax/libs/signature_pad/1.5.3/signature_pad.min.js"></script>
```
You can select a different version at [https://cdnjs.com/libraries/signature_pad](https://cdnjs.com/libraries/signature_pad).

This library is provided as UMD (Universal Module Definition) and ES6 module.

**NOTE** When importing this library in TypeScript, one needs to use the following syntax:
```js
import * as SignaturePad from 'signature_pad';
```
For more info why it's needed, check these 2 issues: [TypeScript#13017](https://github.com/Microsoft/TypeScript/issues/13017) and [rollup#1156](https://github.com/rollup/rollup/issues/1156).

## Usage
### API
``` javascript
var canvas = document.querySelector("canvas");

var signaturePad = new SignaturePad(canvas);

// Returns signature image as data URL (see https://mdn.io/todataurl for the list of possible parameters)
signaturePad.toDataURL(); // save image as PNG
signaturePad.toDataURL("image/jpeg"); // save image as JPEG
signaturePad.toDataURL("image/svg+xml"); // save image as SVG

// Draws signature image from data URL
signaturePad.fromDataURL("data:image/png;base64,iVBORw0K...");

// Returns signature image as an array of point groups
const data = signaturePad.toData();

// Draws signature image from an array of point groups
signaturePad.fromData(data);

// Clears the canvas
signaturePad.clear();

// Returns true if canvas is empty, otherwise returns false
signaturePad.isEmpty();

// Unbinds all event handlers
signaturePad.off();

// Rebinds all event handlers
signaturePad.on();
```

### Options
<dl>
<dt>dotSize</dt>
<dd>(float or function) Radius of a single dot.</dd>
<dt>minWidth</dt>
<dd>(float) Minimum width of a line. Defaults to <code>0.5</code>.</dd>
<dt>maxWidth</dt>
<dd>(float) Maximum width of a line. Defaults to <code>2.5</code>.</dd>
<dt>throttle</dt>
<dd>(integer) Draw the next point at most once per every <code>x</code> milliseconds. Set it to <code>0</code> to turn off throttling. Defaults to <code>16</code>.</dd>
<dt>backgroundColor</dt>
<dd>(string) Color used to clear the background. Can be any color format accepted by <code>context.fillStyle</code>. Defaults to <code>"rgba(0,0,0,0)"</code> (transparent black). Use a non-transparent color e.g. <code>"rgb(255,255,255)"</code> (opaque white) if you'd like to save signatures as JPEG images.</dd>
<dt>penColor</dt>
<dd>(string) Color used to draw the lines. Can be any color format accepted by <code>context.fillStyle</code>. Defaults to <code>"black"</code>.</dd>
<dt>velocityFilterWeight</dt>
<dd>(float) Weight used to modify new velocity based on the previous velocity. Defaults to <code>0.7</code>.</dd>
<dt>onBegin</dt>
<dd>(function) Callback when stroke begin.</dd>
<dt>onEnd</dt>
<dd>(function) Callback when stroke end.</dd>
</dl>

You can set options during initialization:
```javascript
var signaturePad = new SignaturePad(canvas, {
    minWidth: 5,
    maxWidth: 10,
    penColor: "rgb(66, 133, 244)"
});
```
or during runtime:
```javascript
var signaturePad = new SignaturePad(canvas);
signaturePad.minWidth = 5;
signaturePad.maxWidth = 10;
signaturePad.penColor = "rgb(66, 133, 244)";
```


### Tips and tricks
#### Handling high DPI screens
To correctly handle canvas on low and high DPI screens one has to take `devicePixelRatio` into account and scale the canvas accordingly. This scaling is also necessary to properly display signatures loaded via `SignaturePad#fromDataURL`. Here's an example how it can be done:
```javascript
function resizeCanvas() {
    var ratio =  Math.max(window.devicePixelRatio || 1, 1);
    canvas.width = canvas.offsetWidth * ratio;
    canvas.height = canvas.offsetHeight * ratio;
    canvas.getContext("2d").scale(ratio, ratio);
    signaturePad.clear(); // otherwise isEmpty() might return incorrect value
}

window.addEventListener("resize", resizeCanvas);
resizeCanvas();
```
Instead of `resize` event you can listen to screen orientation change, if you're using this library only on mobile devices. You can also throttle the `resize` event - you can find some examples on [this MDN page](https://developer.mozilla.org/en-US/docs/Web/Events/resize).

When you modify width or height of a canvas, it will be automatically cleared by the browser. SignaturePad doesn't know about it by itself, so you can call `signaturePad.clear()` to make sure that `signaturePad.isEmpty()` returns correct value in this case.

This clearing of the canvas by the browser can be annoying, especially on mobile devices e.g. when screen orientation is changed. There are a few workarounds though, e.g. you can [lock screen orientation](https://developer.mozilla.org/en-US/docs/Web/API/Screen/lockOrientation), or read an image from the canvas before resizing it and write the image back after.

#### Handling data URI encoded images on the server side
If you are not familiar with data URI scheme, you can read more about it on [Wikipedia](http://en.wikipedia.org/wiki/Data_URI_scheme).

There are 2 ways you can handle data URI encoded images.

You could simply store it in your database as a string and display it in HTML like this:

``` html
<img src="data:image/png;base64,iVBORw0K..." />
```

but this way has many disadvantages - it's not easy to get image dimensions, you can't manipulate it e.g. to create a thumbnail and it also [has some performance issues on mobile devices](http://www.mobify.com/blog/data-uris-are-slow-on-mobile/).

Thus, more common way is to decode it and store as a file. Here's an example in Ruby:

``` ruby
require "base64"

data_uri = "data:image/png;base64,iVBORw0K..."
encoded_image = data_uri.split(",")[1]
decoded_image = Base64.decode64(encoded_image)
File.open("signature.png", "wb") { |f| f.write(decoded_image) }
```

Here's an example in PHP:

``` php
$data_uri = "data:image/png;base64,iVBORw0K...";
$encoded_image = explode(",", $data_uri)[1];
$decoded_image = base64_decode($encoded_image);
file_put_contents("signature.png", $decoded_image);
```

#### Removing empty space around a signature
If you'd like to remove (trim) empty space around a signature, you can do it on the server side or the client side. On the server side you can use e.g. ImageMagic and its `trim` option: `convert -trim input.jpg output.jpg`. If you don't have access to the server, or just want to trim the image before submitting it to the server, you can do it on the client side as well. There are a few examples how to do it, e.g. [here](https://github.com/szimek/signature_pad/issues/49#issue-29108215) or [here](https://github.com/szimek/signature_pad/issues/49#issuecomment-260976909) and there's also a tiny library [trim-canvas](https://github.com/agilgur5/trim-canvas) that provides this functionality.

#### Drawing over an image
Demo: <https://jsfiddle.net/szimek/d6a78gwq/>

## License
Released under the [MIT License](http://www.opensource.org/licenses/MIT).


CHANGELOG.md
## Changelog
### 2.1.1-alt
*   Compatible with angular2-signaturepad and doesn't break AOT compilation
*   Fixes SVG export colors

### 2.1.1
*   Fixed a bug where default value was applied for throttle when throttle was set to 0. ([mkrause](https://github.com/mkrause) in [#247](https://github.com/szimek/signature_pad/pull/247))

### 2.1.0
*   No changes since 2.1.0-beta.1.

### 2.1.0-beta.1
*   Added throttling. ([@kunukn](https://github.com/kunukn) in [#237](https://github.com/szimek/signature_pad/pull/237))

### 2.0.0
Unfortunately, some breaking changes were introduced in 1.6.0, so to follow the semantic versioning, it's re-released as 2.0.0.
*   Removed support for Bower. If you still need it, use [1.5.3](https://github.com/szimek/signature_pad/releases/tag/v1.5.3) release.
*   Moved `signature_pad.js` and `signature_pad.min.js` files to `dist` folder.
*   Added ES6 version of the library for use with webpack etc.

### 1.6.0 (deprecated in favor of 2.0.0)
*   Added support for returning signature as SVG using `#fromDataURL('image/svg+xml')`. [jackspirou](https://github.com/jackspirou) [mymattcarroll](https://github.com/mymattcarroll) [szimek](https://github.com/szimek)
*   Added `#toData` method that returns data points.
*   Added `#fromData` method that draws signature from data points.
*   Moved `signature_pad.js` and `signature_pad.min.js` files to `dist` folder.

### 1.5.3
*   Fix `touchend` event on touch devices. (#150) [mtomic](https://github.com/mtomic)
*   Fix handling touch events in Egde browser. (#134) [dideldum73](https://github.com/dideldum73)

### 1.5.2
*   Prevent loading an empty string in `fromDataURL`. (#108) [Remo](https://github.com/Remo)
*   Reject points generated by resting hand (better handling of multi touch). (#48 and #57) [jurreantonisse](https://github.com/jurreantonisse)

### 1.5.1
*   Prevent duplicate events on tap in iOS Safari. [PerfectPixel](https://github.com/PerfectPixel)

### 1.5.0
*   Add `on` method that rebinds all event handlers. [Alplob](https://github.com/Alplob)

### 1.4.0
*   Add `off` method that unbinds all event handlers. [Rob-ot](https://github.com/Rob-ot)

### 1.3.6
*   Fix support for Browserify. [chevett](https://github.com/chevett)

### 1.3.5
*   Add support for CommonJS/AMD/UMD.

### 1.3.4
*   Really fix `fromDataURL` on HiDPI screens.

### 1.3.3
*   Fix `fromDataURL` on HiDPI screens.

### 1.3.2
*   Fix `onBegin` and `onEnd` callbacks when passed as options to constructor. [yinsee](https://github.com/yinsee)

### 1.3.1
*   Fix handling touch events on mobile IE. [tocsoft](https://github.com/tocsoft)

### 1.3.0
*   Add `onBegin` and `onEnd` callbacks. [rogerz](https://github.com/rogerz)

### 1.2.4
*   Fix bug where stroke becomes very thin. [mvirkkunen](https://github.com/mvirkkunen)

### 1.2.3
*   Fix `SignaturePad#fromDataURL` on Firefox. [Fr3nzzy](https://github.com/Fr3nzzy)

### 1.2.2
*   Make `SignaturePad#isEmpty` return false after loading an image using `SignaturePad#fromDataURL`. [krisivanov](https://github.com/krisivanov)

### 1.2.1
*   Fixed `SignaturePad#clear()`.

### 1.2.0
*   Add `backgroundColor` option to set custom color of the background on `SignaturePad#clear()`.
*   Rename `color` option to `penColor`.
*   Fix passing arguments to canvas element on `SignaturePad#toDataURL()`.


signature_pad.js
(function (root, factory) {
  if (typeof define === 'function' && define.amd) {
    // AMD. Register as an anonymous module unless amdModuleId is set
    define([], function () {
      return (root['SignaturePad'] = factory());
    });
  } else if (typeof exports === 'object') {
    // Node. Does not work with strict CommonJS, but
    // only CommonJS-like environments that support module.exports,
    // like Node.
    module.exports = factory();
  } else {
    root['SignaturePad'] = factory();
  }
}(this, function () {

/*!
 * Signature Pad v2.1.1
 * https://github.com/szimek/signature_pad
 *
 * Copyright 2017 Szymon Nowak
 * Released under the MIT license
 *
 * The main idea and some parts of the code (e.g. drawing variable width Bézier curve) are taken from:
 * http://corner.squareup.com/2012/07/smoother-signatures.html
 *
 * Implementation of interpolation using cubic Bézier curves is taken from:
 * http://benknowscode.wordpress.com/2012/09/14/path-interpolation-using-cubic-bezier-and-control-point-estimation-in-javascript
 *
 * Algorithm for approximated length of a Bézier curve is taken from:
 * http://www.lemoda.net/maths/bezier-length/index.html
 *
 */
var SignaturePad = (function (document) {
    "use strict";

	function Point(x, y, time) {
	  this.x = x;
	  this.y = y;
	  this.time = time || new Date().getTime();
	}

	Point.prototype.velocityFrom = function (start) {
	  return (this.time !== start.time) ? this.distanceTo(start) / (this.time - start.time) : 1;
	};

	Point.prototype.distanceTo = function (start) {
	  return Math.sqrt(Math.pow(this.x - start.x, 2) + Math.pow(this.y - start.y, 2));
	};

	function Bezier(startPoint, control1, control2, endPoint) {
	  this.startPoint = startPoint;
	  this.control1 = control1;
	  this.control2 = control2;
	  this.endPoint = endPoint;
	}

	// Returns approximated length.
	Bezier.prototype.length = function () {
	  const steps = 10;
	  let length = 0;
	  let px;
	  let py;

	  for (let i = 0; i <= steps; i += 1) {
		const t = i / steps;
		const cx = this._point(
		  t,
		  this.startPoint.x,
		  this.control1.x,
		  this.control2.x,
		  this.endPoint.x
		);
		const cy = this._point(
		  t,
		  this.startPoint.y,
		  this.control1.y,
		  this.control2.y,
		  this.endPoint.y
		);
		if (i > 0) {
		  const xdiff = cx - px;
		  const ydiff = cy - py;
		  length += Math.sqrt((xdiff * xdiff) + (ydiff * ydiff));
		}
		px = cx;
		py = cy;
	  }

	  return length;
	};

	/* eslint-disable no-multi-spaces, space-in-parens */
	Bezier.prototype._point = function (t, start, c1, c2, end) {
	  return (       start * (1.0 - t) * (1.0 - t)  * (1.0 - t))
		   + (3.0 *  c1    * (1.0 - t) * (1.0 - t)  * t)
		   + (3.0 *  c2    * (1.0 - t) * t          * t)
		   + (       end   * t         * t          * t);
	};
	/* eslint-enable no-multi-spaces, space-in-parens */

	function throttle(func, wait, options) {
	  var context, args, result;
	  var timeout = null;
	  var previous = 0;
	  if (!options) options = {};
	  var later = function () {
		previous = options.leading === false ? 0 : Date.now();
		timeout = null;
		result = func.apply(context, args);
		if (!timeout) context = args = null;
	  };
	  return function () {
		var now = Date.now();
		if (!previous && options.leading === false) previous = now;
		var remaining = wait - (now - previous);
		context = this;
		args = arguments;
		if (remaining <= 0 || remaining > wait) {
		  if (timeout) {
			clearTimeout(timeout);
			timeout = null;
		  }
		  previous = now;
		  result = func.apply(context, args);
		  if (!timeout) context = args = null;
		} else if (!timeout && options.trailing !== false) {
		  timeout = setTimeout(later, remaining);
		}
		return result;
	  };
	}


	function SignaturePad(canvas, options) {
	  const self = this;
	  const opts = options || {};

	  this.velocityFilterWeight = opts.velocityFilterWeight || 0.7;
	  this.minWidth = opts.minWidth || 0.5;
	  this.maxWidth = opts.maxWidth || 2.5;
	  this.throttle = 'throttle' in opts ? opts.throttle : 16; // in miliseconds

	  if (this.throttle) {
		this._strokeMoveUpdate = throttle(SignaturePad.prototype._strokeUpdate, this.throttle);
	  } else {
		this._strokeMoveUpdate = SignaturePad.prototype._strokeUpdate;
	  }

	  this.dotSize = opts.dotSize || function () {
		return (this.minWidth + this.maxWidth) / 2;
	  };
	  this.penColor = opts.penColor || 'black';
	  this.backgroundColor = opts.backgroundColor || 'rgba(0,0,0,0)';
	  this.onBegin = opts.onBegin;
	  this.onEnd = opts.onEnd;

	  this._canvas = canvas;
	  this._ctx = canvas.getContext('2d');
	  this.clear();

	  // We need add these inline so they are available to unbind while still having
	  // access to 'self' we could use _.bind but it's not worth adding a dependency.
	  this._handleMouseDown = function (event) {
		if (event.which === 1) {
		  self._mouseButtonDown = true;
		  self._strokeBegin(event);
		}
	  };

	  this._handleMouseMove = function (event) {
		if (self._mouseButtonDown) {
		  self._strokeMoveUpdate(event);
		}
	  };

	  this._handleMouseUp = function (event) {
		if (event.which === 1 && self._mouseButtonDown) {
		  self._mouseButtonDown = false;
		  self._strokeEnd(event);
		}
	  };

	  this._handleTouchStart = function (event) {
		if (event.targetTouches.length === 1) {
		  const touch = event.changedTouches[0];
		  self._strokeBegin(touch);
		}
	  };

	  this._handleTouchMove = function (event) {
		// Prevent scrolling.
		event.preventDefault();

		const touch = event.targetTouches[0];
		self._strokeMoveUpdate(touch);
	  };

	  this._handleTouchEnd = function (event) {
		const wasCanvasTouched = event.target === self._canvas;
		if (wasCanvasTouched) {
		  event.preventDefault();
		  self._strokeEnd(event);
		}
	  };

	  // Enable mouse and touch event handlers
	  this.on();
	}

	// Public methods
	SignaturePad.prototype.clear = function () {
	  const ctx = this._ctx;
	  const canvas = this._canvas;

	  ctx.fillStyle = this.backgroundColor;
	  ctx.clearRect(0, 0, canvas.width, canvas.height);
	  ctx.fillRect(0, 0, canvas.width, canvas.height);

	  this._data = [];
	  this._reset();
	  this._isEmpty = true;
	};

	SignaturePad.prototype.fromDataURL = function (dataUrl) {
	  const image = new Image();
	  const ratio = window.devicePixelRatio || 1;
	  const width = this._canvas.width / ratio;
	  const height = this._canvas.height / ratio;

	  this._reset();
	  image.src = dataUrl;
	  image.onload = () => {
		this._ctx.drawImage(image, 0, 0, width, height);
	  };
	  this._isEmpty = false;
	};

	SignaturePad.prototype.toDataURL = function (type, ...options) {
	  switch (type) {
		case 'image/svg+xml':
		  return this._toSVG();
		default:
		  return this._canvas.toDataURL(type, ...options);
	  }
	};

	SignaturePad.prototype.on = function () {
	  this._handleMouseEvents();
	  this._handleTouchEvents();
	};

	SignaturePad.prototype.off = function () {
	  this._canvas.removeEventListener('mousedown', this._handleMouseDown);
	  this._canvas.removeEventListener('mousemove', this._handleMouseMove);
	  document.removeEventListener('mouseup', this._handleMouseUp);

	  this._canvas.removeEventListener('touchstart', this._handleTouchStart);
	  this._canvas.removeEventListener('touchmove', this._handleTouchMove);
	  this._canvas.removeEventListener('touchend', this._handleTouchEnd);
	};

	SignaturePad.prototype.isEmpty = function () {
	  return this._isEmpty;
	};

	// Private methods
	SignaturePad.prototype._strokeBegin = function (event) {
	  this._data.push([]);
	  this._reset();
	  this._strokeUpdate(event);

	  if (typeof this.onBegin === 'function') {
		this.onBegin(event);
	  }
	};

	SignaturePad.prototype._strokeUpdate = function (event) {
	  const x = event.clientX;
	  const y = event.clientY;

	  const point = this._createPoint(x, y);
	  const { curve, widths } = this._addPoint(point);

	  if (curve && widths) {
		this._drawCurve(curve, widths.start, widths.end);
	  }

	  this._data[this._data.length - 1].push({
		x: point.x,
		y: point.y,
		time: point.time,
    color: this.penColor
	  });
	};

	SignaturePad.prototype._strokeEnd = function (event) {
	  const canDrawCurve = this.points.length > 2;
	  const point = this.points[0];

	  if (!canDrawCurve && point) {
		this._drawDot(point);
	  }

	  if (typeof this.onEnd === 'function') {
		this.onEnd(event);
	  }
	};

	SignaturePad.prototype._handleMouseEvents = function () {
	  this._mouseButtonDown = false;

	  this._canvas.addEventListener('mousedown', this._handleMouseDown);
	  this._canvas.addEventListener('mousemove', this._handleMouseMove);
	  document.addEventListener('mouseup', this._handleMouseUp);
	};

	SignaturePad.prototype._handleTouchEvents = function () {
	  // Pass touch events to canvas element on mobile IE11 and Edge.
	  this._canvas.style.msTouchAction = 'none';
	  this._canvas.style.touchAction = 'none';

	  this._canvas.addEventListener('touchstart', this._handleTouchStart);
	  this._canvas.addEventListener('touchmove', this._handleTouchMove);
	  this._canvas.addEventListener('touchend', this._handleTouchEnd);
	};

	SignaturePad.prototype._reset = function () {
	  this.points = [];
	  this._lastVelocity = 0;
	  this._lastWidth = (this.minWidth + this.maxWidth) / 2;
	  this._ctx.fillStyle = this.penColor;
	};

	SignaturePad.prototype._createPoint = function (x, y, time) {
	  const rect = this._canvas.getBoundingClientRect();

	  return new Point(
		x - rect.left,
		y - rect.top,
		time || new Date().getTime()
	  );
	};

	SignaturePad.prototype._addPoint = function (point) {
	  const points = this.points;
	  let tmp;

	  points.push(point);

	  if (points.length > 2) {
		// To reduce the initial lag make it work with 3 points
		// by copying the first point to the beginning.
		if (points.length === 3) points.unshift(points[0]);

		tmp = this._calculateCurveControlPoints(points[0], points[1], points[2]);
		const c2 = tmp.c2;
		tmp = this._calculateCurveControlPoints(points[1], points[2], points[3]);
		const c3 = tmp.c1;
		const curve = new Bezier(points[1], c2, c3, points[2]);
		const widths = this._calculateCurveWidths(curve);

		// Remove the first element from the list,
		// so that we always have no more than 4 points in points array.
		points.shift();

		return { curve, widths };
	  }

	  return {};
	};

	SignaturePad.prototype._calculateCurveControlPoints = function (s1, s2, s3) {
	  const dx1 = s1.x - s2.x;
	  const dy1 = s1.y - s2.y;
	  const dx2 = s2.x - s3.x;
	  const dy2 = s2.y - s3.y;

	  const m1 = { x: (s1.x + s2.x) / 2.0, y: (s1.y + s2.y) / 2.0 };
	  const m2 = { x: (s2.x + s3.x) / 2.0, y: (s2.y + s3.y) / 2.0 };

	  const l1 = Math.sqrt((dx1 * dx1) + (dy1 * dy1));
	  const l2 = Math.sqrt((dx2 * dx2) + (dy2 * dy2));

	  const dxm = (m1.x - m2.x);
	  const dym = (m1.y - m2.y);

	  const k = l2 / (l1 + l2);
	  const cm = { x: m2.x + (dxm * k), y: m2.y + (dym * k) };

	  const tx = s2.x - cm.x;
	  const ty = s2.y - cm.y;

	  return {
		c1: new Point(m1.x + tx, m1.y + ty),
		c2: new Point(m2.x + tx, m2.y + ty),
	  };
	};

	SignaturePad.prototype._calculateCurveWidths = function (curve) {
	  const startPoint = curve.startPoint;
	  const endPoint = curve.endPoint;
	  const widths = { start: null, end: null };

	  const velocity = (this.velocityFilterWeight * endPoint.velocityFrom(startPoint))
	   + ((1 - this.velocityFilterWeight) * this._lastVelocity);

	  const newWidth = this._strokeWidth(velocity);

	  widths.start = this._lastWidth;
	  widths.end = newWidth;

	  this._lastVelocity = velocity;
	  this._lastWidth = newWidth;

	  return widths;
	};

	SignaturePad.prototype._strokeWidth = function (velocity) {
	  return Math.max(this.maxWidth / (velocity + 1), this.minWidth);
	};

	SignaturePad.prototype._drawPoint = function (x, y, size) {
	  const ctx = this._ctx;

	  ctx.moveTo(x, y);
	  ctx.arc(x, y, size, 0, 2 * Math.PI, false);
	  this._isEmpty = false;
	};

	SignaturePad.prototype._drawCurve = function (curve, startWidth, endWidth) {
	  const ctx = this._ctx;
	  const widthDelta = endWidth - startWidth;
	  const drawSteps = Math.floor(curve.length());

	  ctx.beginPath();

	  for (let i = 0; i < drawSteps; i += 1) {
		// Calculate the Bezier (x, y) coordinate for this step.
		const t = i / drawSteps;
		const tt = t * t;
		const ttt = tt * t;
		const u = 1 - t;
		const uu = u * u;
		const uuu = uu * u;

		let x = uuu * curve.startPoint.x;
		x += 3 * uu * t * curve.control1.x;
		x += 3 * u * tt * curve.control2.x;
		x += ttt * curve.endPoint.x;

		let y = uuu * curve.startPoint.y;
		y += 3 * uu * t * curve.control1.y;
		y += 3 * u * tt * curve.control2.y;
		y += ttt * curve.endPoint.y;

		const width = startWidth + (ttt * widthDelta);
		this._drawPoint(x, y, width);
	  }

	  ctx.closePath();
	  ctx.fill();
	};

	SignaturePad.prototype._drawDot = function (point) {
	  const ctx = this._ctx;
	  const width = (typeof this.dotSize) === 'function' ? this.dotSize() : this.dotSize;

	  ctx.beginPath();
	  this._drawPoint(point.x, point.y, width);
	  ctx.closePath();
	  ctx.fill();
	};

	SignaturePad.prototype._fromData = function (pointGroups, drawCurve, drawDot) {
	  for (let i = 0; i < pointGroups.length; i += 1) {
		const group = pointGroups[i];

		if (group.length > 1) {
		  for (let j = 0; j < group.length; j += 1) {
			const rawPoint = group[j];
			const point = new Point(rawPoint.x, rawPoint.y, rawPoint.time);
			const color = rawPoint.color;

			if (j === 0) {
			  // First point in a group. Nothing to draw yet.
			  this._reset();
			  this._addPoint(point);
			} else if (j !== group.length - 1) {
			  // Middle point in a group.
			  const { curve, widths } = this._addPoint(point);
			  if (curve && widths) {
				drawCurve(curve, widths, color);
			  }
			} else {
			  // Last point in a group. Do nothing.
			}
		  }
		} else {
		  this._reset();
		  const rawPoint = group[0];
		  drawDot(rawPoint);
		}
	  }
	};

	SignaturePad.prototype._toSVG = function () {
	  const pointGroups = this._data;
	  const canvas = this._canvas;
	  const ratio = Math.max(window.devicePixelRatio || 1, 1);
	  const minX = 0;
	  const minY = 0;
	  const maxX = canvas.width / ratio;
	  const maxY = canvas.height / ratio;
	  const svg = document.createElementNS('http://www.w3.org/2000/svg', 'svg');

	  svg.setAttributeNS(null, 'width', canvas.width);
	  svg.setAttributeNS(null, 'height', canvas.height);

	  this._fromData(
		pointGroups,
		(curve, widths, color) => {
		  const path = document.createElement('path');

		  // Need to check curve for NaN values, these pop up when drawing
		  // lines on the canvas that are not continuous. E.g. Sharp corners
		  // or stopping mid-stroke and than continuing without lifting mouse.
		  if (!isNaN(curve.control1.x) &&
			  !isNaN(curve.control1.y) &&
			  !isNaN(curve.control2.x) &&
			  !isNaN(curve.control2.y)) {
			const attr = `M ${curve.startPoint.x.toFixed(3)},${curve.startPoint.y.toFixed(3)} `
					   + `C ${curve.control1.x.toFixed(3)},${curve.control1.y.toFixed(3)} `
					   + `${curve.control2.x.toFixed(3)},${curve.control2.y.toFixed(3)} `
					   + `${curve.endPoint.x.toFixed(3)},${curve.endPoint.y.toFixed(3)}`;

			path.setAttribute('d', attr);
			path.setAttribute('stroke-width', (widths.end * 2.25).toFixed(3));
			path.setAttribute('stroke', color);
			path.setAttribute('fill', 'none');
			path.setAttribute('stroke-linecap', 'round');

			svg.appendChild(path);
		  }
		},
		(rawPoint) => {
		  const circle = document.createElement('circle');
		  const dotSize = (typeof this.dotSize) === 'function' ? this.dotSize() : this.dotSize;
		  circle.setAttribute('r', dotSize);
		  circle.setAttribute('cx', rawPoint.x);
		  circle.setAttribute('cy', rawPoint.y);
		  circle.setAttribute('fill', rawPoint.color);

		  svg.appendChild(circle);
		}
	  );

	  const prefix = 'data:image/svg+xml;base64,';
	  const header = '<svg'
		+ ' xmlns="http://www.w3.org/2000/svg"'
		+ ' xmlns:xlink="http://www.w3.org/1999/xlink"'
		+ ` viewBox="${minX} ${minY} ${maxX} ${maxY}"`
		+ ` width="${maxX}"`
		+ ` height="${maxY}"`
		+ '>';
	  let body = svg.innerHTML;

	  // IE hack for missing innerHTML property on SVGElement
	  if (body === undefined) {
		const dummy = document.createElement('dummy');
		const nodes = svg.childNodes;
		dummy.innerHTML = '';

		for (let i = 0; i < nodes.length; i += 1) {
		  dummy.appendChild(nodes[i].cloneNode(true));
		}

		body = dummy.innerHTML;
	  }

	  const footer = '</svg>';
	  const data = header + body + footer;

	  return prefix + btoa(data);
	};

	SignaturePad.prototype.fromData = function (pointGroups) {
	  this.clear();

	  this._fromData(
		pointGroups,
		(curve, widths) => this._drawCurve(curve, widths.start, widths.end),
		rawPoint => this._drawDot(rawPoint)
	  );
	};

	SignaturePad.prototype.toData = function () {
	  return this._data;
	};


	return SignaturePad;
})(document);

return SignaturePad;

}));
